### DISCUSSION TITLE
PauseAudio and SendDtmf advice needed

### DISCUSSION THREAD

**Discussion by NickDarlowNvm** [sipsorcery-org/sipsorcery#588] [General]
Referencing the SIPCallServer example I have created a WPF application that calls Linphone and plays music. The call is initiated on pressing a Start button and ended on clicking a Stop button.

When I press a Dtmf button I want to stop playing the music, play a DTMF tone and resume playing the music. Unfortunately this isn't working; the music does not stop and no DTMF tone is heard. I expect I've forgotten something but I can't spot what. Please can someone point me in the right direction?

```
private async void Start(object sender, RoutedEventArgs e)
{
    _transport = new SIPTransport();
    _agent = new SIPUserAgent(_transport, null);

    _codecs = new List<AudioCodecsEnum> { AudioCodecsEnum.PCMU, AudioCodecsEnum.PCMA };
    _audioExtrasSource = new AudioExtrasSource(new AudioEncoder(), new AudioSourceOptions { AudioSource = AudioSourcesEnum.Music });
    _audioExtrasSource.RestrictFormats(format => _codecs.Contains(format.Codec));

    _session = new VoIPMediaSession(new MediaEndPoints { AudioSource = _audioExtrasSource });
    _session.AcceptRtpFromAny = true;

    if (await _agent.Call(Destination, null, null, _session))
        await _session.Start();
}

private async void Dtmf(object sender, RoutedEventArgs e)
{
    await _session.AudioExtrasSource.PauseAudio();
    await _agent.SendDtmf(0x5);
    await Task.Delay(1000);
    await _session.AudioExtrasSource.ResumeAudio();
}

private void Stop(object sender, RoutedEventArgs e)
{
    _agent.Hangup();
}
```

**Comment by sipsorcery** [comment]
The first part of the problem, with the music not pausing, is because the [Pause](https://github.com/sipsorcery-org/sipsorcery/blob/8f38a9c07a7910eda1d378e7e35b73bb7f459001/src/app/Media/Sources/AudioExtrasSource.cs#L227) method doesn't do anything other than setting a variable. I'd never really intended for the sipsorcery library to start dealing with media and only added classes like `AudioExtraSource` as examples. As is generally the case, the best laid plans go awry and the example classes have become relied upon and their deficiencies are being laid bare. An easy alternative would be to set the source to `AudioSourcesEnum.Silence`. A better alternative would be to fix the logic so that pause and resume do what they are expected to do and stop and resume sending audio samples.

The DTMF tones are a separate issue. The library only supports DTMF via [RTP Events](https://github.com/sipsorcery-org/sipsorcery/blob/8f38a9c07a7910eda1d378e7e35b73bb7f459001/src/net/RTP/RTPSession.cs#L1912) (RFC2833) packets. If you're not hearing DTMF tones it may be because the remote peer doesn't support that mechanism or is configured for a different one.

**Reply by budimulyawan** [reply]
how do i set silence? i tried  voipMediaSession.AudioExtrasSource.SetSource(AudioSourcesEnum.Silence); but it seems not working i.e the sine wave is still playing

**Comment by NickDarlowNvm** [comment]
Thanks for your response Aaron and for your efforts with Sip Sorcery. The code base I am currently working with uses pjsip with a C# wrapper as part of a test library and I was exploring potential alternatives.

The PlayDtmf example worked for me but uses WindowsAudioEndPoint. I was hoping to avoid any Windows specific library as the tests will run on Windows and Linux hosts. Perhaps use of WindowsAudioEndPoint is integral? I tried calling AudioExtrasSource.SetSource and it didn’t do anything once the audio had started unless I had created the session using WindowsAudioEndPoint.

I take your point on contributing to the library but am not sure I can commit the time. Thanks once again for your efforts.

**Reply by budimulyawan** [reply]
did you managed to pause the music? how did u do it? i got same  AudioExtrasSource.SetSource that is not working once the audio started
